_source/_posts/2020-12-23-mulesoft-salesforce-integration.md

```diff
@@ -0,0 +1,650 @@
+
+HTTPResponse resp = http.send(req);
+
+if(resp.getBody() != null)
+
+{
+
+Map<String, String> values = (Map<String, String>)JSON.deserialize(resp.getBody(), Map<String, String>.class);
+
+
+sub = values.get('sub');
+
+name = values.get('name');
+
+email = values.get('email');
+
+userName = values.get('preferred_username');
+
+firstName = values.get('given_name');
+
+lastname = values.get('family_name');
+
+}
+
+return Auth.AuthToken.getAccessToken('0SO2v000000XjXAGA0', 'Open ID Connect'); //change the first parameter to the Auth. Provider ID obtained earlier.
+
+}
+
+public PageReference fetch_data() {
+
+HttpRequest req = new HttpRequest();
+
+Http http = new Http();
+
+String url = ' http://protectedmulesoftapi.us-e2.cloudhub.io/contacts'; //change this to your Mulesoft API proxy endpoint you've created earlier
+
+req.setEndpoint(url);
+
+req.setMethod('GET');
+
+req.setHeader('Accept', 'application/json');
+
+req.setHeader('Authorization', 'Bearer '+ Auth.AuthToken.getAccessToken('0SO2v000000XjXAGA0', 'Open ID Connect')); //change the first parameter to the Auth. Provider ID obtained earlier.
+
+HTTPResponse resp = http.send(req);
+
+system.debug('INSIDE CALLOUT:'+resp.getBody());
+
+if(resp.getBody() != null)
+
+{
+
+Map<String, String> values = (Map<String, String>)JSON.deserialize(resp.getBody(), Map<String, String>.class);
+
+String append ='';
+
+append = values.get('FirstName') + ':' + values.get('LastName') + ':' + values.get('Email') + ':' + values.get('Company');
+
+callOut = append;
+
+}
+
+return null;
+
+}
+
+global boolean canCreateUser(Auth.UserData data) {
+
+//TODO: Check whether we want to allow the creation of a user with this data
+
+return true;
+
+}
+
+global User createUser(Id portalId, Auth.UserData data){
+
+if(!canCreateUser(data)) {
+
+//Returning null or throwing an exception fails the SSO flow
+
+return null;
+
+}
+
+//The user is authorized, so create their Salesforce user
+
+User u = new User();
+
+Profile p = [SELECT Id FROM profile WHERE name='Standard Platform User'];
+
+//TODO: Customize the username. Also, check that the username doesn't already exist and
+
+//possibly ensure there are enough org licenses to create a user. Must be 80 characters
+
+//or less.
+
+u.username = data.username;
+
+u.email = data.email;
+
+u.lastName = data.lastName;
+
+u.firstName = data.firstName;
+
+String alias = data.username;
+
+//Alias must be 8 characters or less
+
+if(alias.length() > 8) {
+
+alias = alias.substring(0, 8);
+
+}
+
+u.alias = alias;
+
+u.languagelocalekey = 'en_US';
+
+u.localesidkey = 'en_US';
+
+u.emailEncodingKey = 'UTF-8';
+
+u.timeZoneSidKey = 'America/Los_Angeles';
+
+u.profileId = p.Id;
+
+insert u;
+
+return u;
+
+}
+
+global void updateUser(Id userId, Id portalId, Auth.UserData data){
+
+User u = new User(id=userId);
+
+u.email = data.email;
+
+u.lastName = data.lastName;
+
+u.firstName = data.firstName;
+
+}
+
+```
+
```

Comment:
```suggestion

```java
    public String accessToken;
    public String callOut{get;set;}
    public String sub{get;set;}
    public String name{get;set;}
    public String email{get;set;}
    public String firstName{get;set;}
    public String lastName{get;set;}
    public String userName{get;set;}

    public String getAccessToken()
    {
        HttpRequest req = new HttpRequest();
        Http http = new Http();
        String url = ' https://<yourOktaDomain>/oauth2/default/v1/userinfo'; //change this to your Okta developer instance userinfo url
        req.setEndpoint(url);
        req.setMethod('GET');
        req.setHeader('Accept', 'application/json');
        req.setHeader('Authorization', 'Bearer '+ Auth.AuthToken.getAccessToken('0SO2v000000XjXAGA0', 'Open ID Connect'));//change the first parameter to the Auth. Provider ID obtained earlier.
        HTTPResponse resp = http.send(req);
        if(resp.getBody() != null)
        {
            Map<String, String> values = (Map<String, String>)JSON.deserialize(resp.getBody(), Map<String, String>.class);
            sub = values.get('sub');
            name = values.get('name');
            email = values.get('email');
            userName = values.get('preferred_username');
            firstName = values.get('given_name');
            lastname = values.get('family_name');
        }
        return Auth.AuthToken.getAccessToken('0SO2v000000XjXAGA0', 'Open ID Connect'); //change the first parameter to the Auth. Provider ID obtained earlier.
    }
    public PageReference fetch_data() {
        HttpRequest req = new HttpRequest();
        Http http = new Http();
        String url = ' http://protectedmulesoftapi.us-e2.cloudhub.io/contacts'; //change this to your Mulesoft API proxy endpoint you've created earlier
        req.setEndpoint(url);
        req.setMethod('GET');
        req.setHeader('Accept', 'application/json');
        req.setHeader('Authorization', 'Bearer '+ Auth.AuthToken.getAccessToken('0SO2v000000XjXAGA0', 'Open ID Connect')); //change the first parameter to the Auth. Provider ID obtained earlier.
        HTTPResponse resp = http.send(req);
        system.debug('INSIDE CALLOUT:'+resp.getBody());
        if(resp.getBody() != null)
        {
            Map<String, String> values = (Map<String, String>)JSON.deserialize(resp.getBody(), Map<String, String>.class);
            String append ='';
            append = values.get('FirstName') + ':' + values.get('LastName') + ':' + values.get('Email') + ':' + values.get('Company');
            callOut = append;
        }
        return null;
    }
    global boolean canCreateUser(Auth.UserData data) {
//TODO: Check whether we want to allow the creation of a user with this data
        return true;
    }
    global User createUser(Id portalId, Auth.UserData data){
        if(!canCreateUser(data)) {
//Returning null or throwing an exception fails the SSO flow
            return null;
        }
//The user is authorized, so create their Salesforce user
        User u = new User();
        Profile p = [SELECT Id FROM profile WHERE name='Standard Platform User'];
//TODO: Customize the username. Also, check that the username doesn't already exist and
//possibly ensure there are enough org licenses to create a user. Must be 80 characters
//or less.
        u.username = data.username;
        u.email = data.email;
        u.lastName = data.lastName;
        u.firstName = data.firstName;
        String alias = data.username;
//Alias must be 8 characters or less
        if(alias.length() > 8) {
            alias = alias.substring(0, 8);
        }
        u.alias = alias;
        u.languagelocalekey = 'en_US';
        u.localesidkey = 'en_US';
        u.emailEncodingKey = 'UTF-8';
        u.timeZoneSidKey = 'America/Los_Angeles';
        u.profileId = p.Id;
        insert u;
        return u;
    }
    global void updateUser(Id userId, Id portalId, Auth.UserData data){
        User u = new User(id=userId);
        u.email = data.email;
        u.lastName = data.lastName;
        u.firstName = data.firstName;
    }
}
```

```

Reply:
This isn't actually Java, but I tried to reformat the code, more tweaks could be done, i'm mostly guessing